pkg/operator/customroute/custom_route_controller.go

```diff
@@ -0,0 +1,276 @@
+			Name:      componentRouteName,
+		},
+		targetRoute:    targetRoute.DeepCopy(),
+		consumingUsers: consumingUsers,
+		ingressLister:  ingressInformer.Lister(),
+		ingressClient:  ingressClient,
+		routeLister:    routeInformer.Lister(),
+		routeClient:    routeClient,
+		secretLister:   kubeInformersForNamespaces.SecretLister(),
+		operatorClient: operatorClient,
+		resourceSyncer: resourceSyncer,
+	}
+
+	return factory.New().
+		WithInformers(
+			ingressInformer.Informer(),
+			routeInformer.Informer(),
+			kubeInformersForNamespaces.InformersFor("openshift-config").Core().V1().Secrets().Informer(),
+			kubeInformersForNamespaces.InformersFor(destSecretNamespace).Core().V1().Secrets().Informer(),
+		).
+		WithSyncDegradedOnError(operatorClient).
+		WithSync(controller.sync).
+		ResyncEvery(time.Minute).
+		ToController("CustomRouteController", eventRecorder.WithComponentSuffix("custom-route-controller"))
+}
+
+func (c *customRouteController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	ingressConfig, err := c.ingressLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+
+	ingressConfigCopy := ingressConfig.DeepCopy()
+
+	errors := c.syncResources(ctx, syncCtx, ingressConfigCopy)
+	return c.updateIngressConfigStatus(ctx, ingressConfigCopy, errors)
+}
+
+func (c *customRouteController) syncResources(ctx context.Context, syncCtx factory.SyncContext, ingressConfig *configv1.Ingress) []error {
+	expectedRoute, secretName, errors := c.getOAuthRouteAndSecretName(ingressConfig)
+	if errors != nil {
+		return errors
+	}
+
+	if _, _, err := resourceapply.ApplyRoute(ctx, c.routeClient, syncCtx.Recorder(), expectedRoute); err != nil {
+		return []error{err}
+	}
+
+	if err := c.syncSecret(secretName); err != nil {
+		return []error{err}
+	}
+
+	return []error{}
+}
+
+func (c *customRouteController) getOAuthRouteAndSecretName(ingressConfig *configv1.Ingress) (*routev1.Route, string, []error) {
+	route := c.targetRoute
+	// set defaults
+	route.Spec.Host = route.ObjectMeta.Name + "." + ingressConfig.Spec.Domain // mimic the behavior of subdomain
+	secretName := ""
+
+	// check if a user is overriding route defaults
+	if componentRoute := routecomponenthelpers.GetComponentRouteSpec(ingressConfig, route.ObjectMeta.Namespace, route.ObjectMeta.Name); componentRoute != nil {
+		var errors []error
+		// Check if the provided secret is valid
+		secretName = componentRoute.ServingCertKeyPairSecret.Name
+		if err := c.validateCustomTLSSecret(secretName); err != nil {
+			errors = append(errors, err)
+		}
+
+		// Check if the provided hostname is valid
+		hostname := string(componentRoute.Hostname)
+		if _, err := url.Parse(hostname); err != nil {
+			errors = append(errors, err)
+		}
+
+		if errors != nil {
+			return nil, "", errors
+		}
+
+		route.Spec.Host = hostname
+	}
+
+	return route, secretName, nil
+}
+
+func (c *customRouteController) validateCustomTLSSecret(secretName string) error {
+	if secretName != "" {
```

Comment:
```golang
if len(secretName) == 0 {
    return nil
}
... the rest here...
```

Reply:
Just to educate me - what's the difference between those two?

Reply:
Fixed.

Reply:
The difference is that you don't have to deal with the indentation for the rest of the function, therefore it's considered more clean.